pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +499,99 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) OwnedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) TriggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+func (u *userTSDB) RecomputeOwnedSeries(shardSize int, reason string, logger log.Logger) {
+	// We need to recompute owned series, ie. how many series in this user's Head are owned by this ingester (according to
+	// current token ranges), and updates both ownedSeries and ownedSeriesShardSize.
+
+	start := time.Now()
+	repeats := 0
+
+	ownedPrev, shardSizePrev := u.OwnedSeriesAndShards()
+
+	var ownedNew int
+	for {
```

Comment:
This could potentially lead to an infinite loop, if new series are added continuously. I think it's a bad idea. I would limit the max number of attempts.

Reply:
I don't think it can lead to infinite loop, because eventually user will hit the limit :) But I agree with your proposal to allow small difference (say up to 100 series). We can also limit the attempts, set whatever value we have, but also  report error to retry later. WDYT?

Reply:
As a best practice we should never have a potentially infinite loop. I don't want to even reason whether could be infinite or not because of "other conditions outside of the loop". However, I've seen you adding a max retries, so LGTM.